### PR TITLE
feat(react-grid): add scroll to row in virtual table

### LINKED ISSUE #436: Scroll to row feature?
Hi,

When a user causes a row to be added to the grid through a dialogue etc, is there a way I can scroll the grid to show the new row?

Thanks,

Dan

### PR DESCRIPTION
**Pull request by LazyLahtak**
Fixes #436

### CODE REVIEW

File: packages/dx-grid-core/src/plugins/virtual-table/helpers.test.ts [hunk 1]
```diff
@@ -170,4 +170,18 @@ describe('#checkColumnWidths', () => {
...
+
+  describe('#calculateScrollHeight', () => {
+    const rowHeight = 100;
+
+    it('should work', () => {
+      expect(calculateScrollHeight(rowHeight, 10))
+        .toEqual(1000);
+    });
+
+    it('should return "undefined" if index not defined', () => {
```
Comment (AryamnovEugeniy) [inline packages/dx-grid-core/src/plugins/virtual-table/helpers.test.ts]: ... if index is not...

File: packages/dx-grid-core/src/plugins/virtual-table/helpers.test.ts [hunk 2]
```diff
@@ -170,4 +170,18 @@ describe('#checkColumnWidths', () => {
...
+      expect(calculateScrollHeight(rowHeight, undefined))
+        .toEqual(undefined);
+    });
```
Comment (AryamnovEugeniy) [inline packages/dx-grid-core/src/plugins/virtual-table/helpers.test.ts]: Should we add a test with a negative index?

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 3]
```diff
@@ -37,3 +37,11 @@ The Grid also supports column virtualization, which is demonstrated in the follo
 If the Grid should have the same size as the container element, set the `VirtualTable` plugin's `height` property to "auto" and the Grid root element's style setting to `height: 100%`.
 
 .embedded-demo({ "path": "grid-virtual-scrolling/stretching-to-parent-element", "showThemeSelector": true })
+
+## Scroll to Row
+
+The virtual table can be scrolled programmatically, through the `scrollToIndex` and `scrollToId` methods. To call these, you should use `VirtualTable` plugin's reference. The following demo demonstrates how to do that:
```
Comment (AryamnovEugeniy) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]: We should add a link to VirtualTable's Methods section

File: packages/dx-react-core/src/sizer.tsx [hunk 4]
```diff
@@ -126,11 +126,22 @@ export class Sizer extends React.PureComponent<SizerProps> {
     this.contractTrigger.appendChild(this.contractNotifier);
   }
 
+  componentDidUpdate() {
+    // We can scroll the VirtualTable manually only by changing
+    // containter's (rootNode) scrollTop property.
+    // Viewport changes its own properties automatically.
+    const { scrollHeight } = this.props;
```
Comment (ushkal) [inline packages/dx-react-core/src/sizer.tsx]: better rename it to `scrollTop` because `scrollHeight` is confusing

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 5]
```diff
@@ -38,6 +38,15 @@ stubRowComponent | ComponentType&lt;[Table.RowProps](#tablerowprops)&gt; | | A c
 stubCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub table cell if the cell value is not provided.
 stubHeaderCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub header cell if the cell value is not provided.
 messages? | [Table.LocalizationMessages](table.md#localization-messages) | | An object that specifies the localization messages.
+onRowIndexChange? | (index: number) => void | | Handles top row index changes.
```
Comment (ushkal) [inline packages/dx-react-grid/docs/reference/virtual-table.md]: We should consider combining these events into one with args = { index, rowId } because their purpose is the same

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 6]
```diff
@@ -38,6 +38,15 @@ stubRowComponent | ComponentType&lt;[Table.RowProps](#tablerowprops)&gt; | | A c
...
+onRowIdChange? | (id: number &#124; string) => void | | Handles top row ID changes.
+
+## Methods
+
+Name | Type | Description
+-----|------|------------
+scrollToIndex | (index: number) => void | A function, that scroll table to specified index.
```
Comment (ushkal) [inline packages/dx-react-grid/docs/reference/virtual-table.md]: The same goes for scrollTo.. methods - we can replace them with `scrollToRow`

File: packages/dx-react-grid-demos/src/demo-sources/grid-virtual-scrolling/scroll-to-row.partial.jsxt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  const [id, setId] = useState(0);
+  const vtRef = useRef();
+
+  const topRowChange = ({ id: newId, index: newIndex }) => {
+    setIndex(newIndex);
+    setId(newId);
+  };
+
+  const goToRow = (row) => {
+    vtRef.current.scrollToRow(row);
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid-demos/src/demo-sources/grid-virtual-scrolling/scroll-to-row.partial.jsxt]: This method calls a `setState()` function. Why we go our from React basic concepts `controlled/uncontrolled` modes via properties. Why we should call the `setState()` method manually? I think it is not a good idea.

Can we use a pair of properties `topRow` and `onTopRowChange`?

Comment (LazyLahtak) [reply packages/dx-react-grid-demos/src/demo-sources/grid-virtual-scrolling/scroll-to-row.partial.jsxt]: We can't use the controlled mode, because the index matches with an interval of height. We cannot identify scroll position by index correctly (scroll is blocked).
We decided to add method, for manual scrolling.

**Comment by MaximKudriavtsev** [comment]
image
looks strange

**Comment by LazyLahtak** [comment]
FPR

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 8]
```diff
@@ -38,6 +38,14 @@ If the Grid should have the same size as the container element, set the `Virtual
 
 .embedded-demo({ "path": "grid-virtual-scrolling/stretching-to-parent-element", "showThemeSelector": true })
 
+## Scroll to Row
+
+The virtual table can be scrolled programmatically, through the `scrollToRow` methods. To call these, you should use `VirtualTable` plugin's reference. The following demo demonstrates how scroll table to added row :
```
Comment (MaximKudriavtsev) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]: `The following demo demonstrates how scroll table to added row` ->
`The following demo demonstrates how scroll table to new and added rows`

Comment (LazyLahtak) [reply packages/dx-react-grid/docs/guides/virtual-scrolling.md]: Corrected

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 9]
```diff
@@ -38,6 +38,14 @@ If the Grid should have the same size as the container element, set the `Virtual
 
 .embedded-demo({ "path": "grid-virtual-scrolling/stretching-to-parent-element", "showThemeSelector": true })
 
+## Scroll to Row
+
+The virtual table can be scrolled programmatically, through the `scrollToRow` methods. To call these, you should use `VirtualTable` plugin's ref property. The following demo demonstrates how scroll table to new and added rows:
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
To scroll the table to a particular row, call the `scrollToRow` method. It accepts a row ID. To call the method, you need the `VirtualTable` plugin's ref.

In the following demo, the `scrollToRow` method is used to scroll the table to a new or saved row. When you add a new row, it is added to the table's top, and the table is scrolled to it. When you save the row, its position is changed according to sorting, and the table is scrolled to that position.
```

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 10]
```diff
@@ -38,6 +38,14 @@ If the Grid should have the same size as the container element, set the `Virtual
...
+
+.embedded-demo({ "path": "grid-virtual-scrolling/scroll-to-row", "showThemeSelector": true })
+
+Note that the virtual table with Lazy Loading doesn't allow to scroll to row's id.
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
NOTE: Scrolling to a row does not work with lazy loading. This feature requires the IDs of all rows, but in lazy loading mode, the Grid loads rows in parts.
```

Comment (RomanTsukanov) [reply packages/dx-react-grid/docs/guides/virtual-scrolling.md]: Please make the phrase "lazy loading" a link to [this article](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/lazy-loading/#react-grid---virtual-scrolling-with-remote-data-lazy-loading).

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 11]
```diff
@@ -38,6 +38,13 @@ stubRowComponent | ComponentType&lt;[Table.RowProps](#tablerowprops)&gt; | | A c
 stubCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub table cell if the cell value is not provided.
 stubHeaderCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub header cell if the cell value is not provided.
 messages? | [Table.LocalizationMessages](table.md#localization-messages) | | An object that specifies the localization messages.
+onTopRowChange? | (rowId: number &#124; string) => void | | Handles top row index and id changes.
+
+## Methods
+
+Name | Type | Description
+-----|------|------------
+scrollToRow | (rowId: number &#124; string) => void | A function, that scroll table to specified row.
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/reference/virtual-table.md]:
```suggestion
scrollToRow | (rowId: number &#124; string) => void | Scrolls table to a row with the specified ID.
```

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 12]
```diff
@@ -38,6 +38,13 @@ stubRowComponent | ComponentType&lt;[Table.RowProps](#tablerowprops)&gt; | | A c
 stubCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub table cell if the cell value is not provided.
 stubHeaderCellComponent | ComponentType&lt;[Table.CellProps](table.md#tablecellprops)&gt; | | A component that renders a stub header cell if the cell value is not provided.
 messages? | [Table.LocalizationMessages](table.md#localization-messages) | | An object that specifies the localization messages.
+onTopRowChange? | (rowId: number &#124; string) => void | | Handles top row index and id changes.
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/reference/virtual-table.md]:
```suggestion
onTopRowChange? | (rowId: number &#124; string) => void | | Handles a change of the top row.
```

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 13]
```diff
@@ -77,6 +84,8 @@ Field | Type | Description
 COLUMN_TYPE | symbol | The data column type's indentifier.
 ROW_TYPE | symbol | The data row type's indentifier.
 NODATA_ROW_TYPE | symbol | The nodata row type's indentifier.
+TOP_POSITION | symbol | The top table position for scroll.
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/reference/virtual-table.md]:
```suggestion
TOP_POSITION | symbol | The top position of the table. Used in scrolling.
```

File: packages/dx-react-grid/docs/reference/virtual-table.md [hunk 14]
```diff
@@ -77,6 +84,8 @@ Field | Type | Description
...
+BOTTOM_POSITION | symbol | The bottom table position for scroll.
```
Comment (RomanTsukanov) [inline packages/dx-react-grid/docs/reference/virtual-table.md]:
```suggestion
BOTTOM_POSITION | symbol | The bottom position of the table. Used in scrolling.
```

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 15]
```diff
@@ -38,6 +38,16 @@ If the Grid should have the same size as the container element, set the `Virtual
 
 .embedded-demo({ "path": "grid-virtual-scrolling/stretching-to-parent-element", "showThemeSelector": true })
 
+## Scroll to Row
+
+To scroll the table to a particular row, call the `scrollToRow` method. It accepts a row ID. To call the method, you need the `VirtualTable` plugin's ref.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]: Perhaps:

"To scroll the table to a particular row, call the `scrollToRow` method and pass the row ID as its parameter"?

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 16]
```diff
@@ -38,6 +38,16 @@ If the Grid should have the same size as the container element, set the `Virtual
...
+
+In the following demo, the `scrollToRow` method is used to scroll the table to a new or saved row. When you add a new row, it is added to the table's top, and the table is scrolled to it. When you save the row, its position is changed according to sorting, and the table is scrolled to that position.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
In the following demo, the `scrollToRow` method is used to scroll the table to a new or saved row. When you add a new row, it is added to the top of the table, and the table is scrolled to it. When you save the row, its position is changed according to sorting, and the table is scrolled to that position.
```

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md [hunk 17]
```diff
@@ -38,6 +38,16 @@ If the Grid should have the same size as the container element, set the `Virtual
...
+
+.embedded-demo({ "path": "grid-virtual-scrolling/scroll-to-row", "showThemeSelector": true })
+
+NOTE: Scrolling to a row does not work with [lazy loading](./lazy-loading.md/#react-grid---virtual-scrolling-with-remote-data-lazy-loading). This feature requires the IDs of all rows, but in lazy loading mode, the Grid loads rows in parts.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
NOTE: Scrolling to a row cannot be used with [lazy loading](./lazy-loading.md/#react-grid---virtual-scrolling-with-remote-data-lazy-loading). This is because the Grid loads rows in parts in lazy loading mode and this feature requires all the row IDs.
```

Comment (RomanTsukanov) [reply packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
NOTE: Scrolling to a row cannot be used with [lazy loading](./lazy-loading.md/#react-grid---virtual-scrolling-with-remote-data-lazy-loading). This is because the Grid loads rows in parts in lazy loading mode, and scrolling to a row requires all the row IDs.
```

File: packages/dx-react-grid/docs/guides/virtual-scrolling.md (hunk 15 above)
Comment (RomanTsukanov) [reply packages/dx-react-grid/docs/guides/virtual-scrolling.md]:
```suggestion
To scroll the table to a particular row, call the `scrollToRow` method and pass the row ID as its parameter. To call the method, you need the `VirtualTable` plugin's ref.
```

**Comment by stevebaxter** [comment]
This is good news!! When might this be released?

**Comment by LazyLahtak** [comment]
@stevebaxter

The next release is scheduled for the next week.

**Comment by stevebaxter** [comment]
Finally got a chance to try this out, unfortunately we've hit two problems:
1. This scrolls the selected rowID to the top rather than scrolling it into view if needed. Is there any way to only scroll if needed?
2. This doesn't seem to work where the table has a mix of row heights (our original issue).
Is there any way around either of these issues?

**Comment by LazyLahtak** [comment]
Hi @stevebaxter,

We will consider implementing this feature in future releases. You can follow [this](https://github.com/DevExpress/devextreme-reactive/issues/2757) topic to watch for our updates.

Also, we are working on the ability to support rows with different heights.
